pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Map;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.config.provider.PinotClusterConfigChangeListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Used to throttle the total concurrent index rebuilds that can happen on a given Pinot server.
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * If _maxPreprocessConcurrency is <= 0, it means that the cluster is not configured to limit the number of
+   * index rebuilds that can be executed concurrently. This class sets the semaphore permit to be a large value
+   * in this scenario.
+   */
+  private int _maxPreprocessConcurrency;
+  private boolean _relaxThrottling;
+  private volatile AdjustableSemaphore _semaphore;
+
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, relaxThrottling);
+    if (maxPreprocessConcurrency <= 0) {
+      int defaultSegmentPreprocessParallelism =
+          Integer.parseInt(CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM);
+      LOGGER.warn("Max preprocess concurrency is negative: {}, resetting to {}", maxPreprocessConcurrency,
+          defaultSegmentPreprocessParallelism);
+      maxPreprocessConcurrency = defaultSegmentPreprocessParallelism;
+    }
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _relaxThrottling = relaxThrottling;
+
+    int relaxThrottlingThreshold = Math.max(_maxPreprocessConcurrency,
+        CommonConstants.Helix.DEFAULT_MAX_SEGMENT_PREPROCESS_PARALLELISM_BEFORE_SERVING_QUERIES);
```

Comment:
We might want a even higher number so that it doesn't throttle. Also consider making it configurable in case we really need to throttle during server restart

Reply:
technically we'll be limited by STATE_TRANSITION.maxThreads, and maybe an additional one from USER_DEFINE_MSG handling for refresh/reload. I've made the default higher though, in case some environments have increased STATE_TRANSITION.maxThreads

I've also made this configurable, but haven't made it configurable via ZK cluster configs. Reason for that is this is only relevant during server startup prior to enabling queries on the server. I think in this case requiring server restart to pick up this config is fine? Let me know if you think otherwise though, and I'll make the changes to allow setting this via ZK cluster configs in that case.

Reply:
Since the one during query serving is a ZK cluster config, I think we can just make the one before query serving also a cluster config. It might be useful when we find server starting up too slow, and adjust the value without restarting the server.

Reply:
good point, made the changes